library/core/src/ptr/iter.rs

```diff
@@ -0,0 +1,513 @@
+                        } else {
+                            self.end = self.start;
+                            None
+                        }
+                    }
+                }
+            }
+
+            fn advance_back_by(&mut self, n: usize) -> Result<(), usize> {
+                match self.size_hint().1 {
+                    None => {
+                        // T is zero sized. Advancing does nothing.
+                        Ok(())
+                    }
+                    Some(len) => {
+                        if n <= len {
+                            // Advance leaving `len - n` elements in the
+                            // iterator.
+                            //
+                            // SAFETY: this pointer remains in between `start`
+                            // and `end` which is in bounds of the same
+                            // allocated object according to the preconditions
+                            // of `new`.
+                            self.end = unsafe { self.start.add(len - n) };
+                            Ok(())
+                        } else {
+                            // Advance too far.
+                            self.end = self.start;
+                            Err(len)
+                        }
+                    }
+                }
+            }
+        }
+
+        #[unstable(feature = "ptr_iter", issue = "none")]
+        impl<T> FusedIterator for $ty<T> {}
+
+        #[unstable(feature = "trusted_len", issue = "37572")]
+        unsafe impl<T> TrustedLen for $ty<T> {}
+
+        #[doc(hidden)]
+        #[unstable(feature = "trusted_random_access", issue = "none")]
+        unsafe impl<T> TrustedRandomAccess for $ty<T> {}
+
+        #[doc(hidden)]
+        #[unstable(feature = "trusted_random_access", issue = "none")]
+        unsafe impl<T> TrustedRandomAccessNoCoerce for $ty<T> {
+            const MAY_HAVE_SIDE_EFFECT: bool = false;
+        }
+
+        // not Copy -- see #27186
+        #[unstable(feature = "ptr_iter", issue = "none")]
+        impl<T> Clone for $ty<T> {
+            fn clone(&self) -> Self {
+                $ty {
+                    start: self.start,
+                    end: self.end,
+                }
+            }
+        }
+
+        #[unstable(feature = "ptr_iter", issue = "none")]
+        impl<T> PartialEq for $ty<T> {
+            fn eq(&self, other: &Self) -> bool {
+                self.start == other.start && self.end == other.end
+            }
+        }
+
+        #[unstable(feature = "ptr_iter", issue = "none")]
+        impl<T> Eq for $ty<T> {}
+
+        #[unstable(feature = "ptr_iter", issue = "none")]
+        impl<T> Hash for $ty<T> {
+            fn hash<H: Hasher>(&self, state: &mut H) {
+                self.start.hash(state);
+                self.end.hash(state);
+            }
+        }
+
+        #[unstable(feature = "ptr_iter", issue = "none")]
+        impl<T> Debug for $ty<T> {
```

Comment:
Not a problem, but I was surprised: These `PartialEq+Eq+Hash+Debug` look like they match what `derive` would do?

Reply:
Derive would've done `impl<T: Debug> Debug for $ty<T>`.

Reply:
Oh, right of course 🤦

I think I brainfarted and thought that the generic was `*const T`, where it would be fine (like it was in `Range<*const T>`) but it's not.  Sorry for noise.